### PR TITLE
Add ERC721EnumerableComponent extension

### LINKED ISSUE #722: Migrate ERC721 Enumerate extension
**🧐 Motivation**

To be feature partiy with the Cairo 0 version, having a ERC721 Enumerable extension proposition would be interesting.

**📝 Details**

Here are 2 repo which did an implementation:
- https://github.com/carbonable-labs/cairo-erc-721/tree/main/src/extensions/enumerable
- https://github.com/Seraph-Labs/Cairo-Contracts/blob/main/seraphlabs/tokens/src/erc721/extensions/enumerable.cairo

### PR DESCRIPTION
**Pull request by andrew-fleming**
Fixes #722.

Deps:
- #978 

#### PR Checklist

- [x] Tests
- [x] Documentation
- [x] Added entry to CHANGELOG.md 
- [ ] Tried the feature on a public network

### CODE REVIEW

File: CHANGELOG.md
Comment (ericnordelo) [inline CHANGELOG.md:13]: Lets not forget to add the entry to the changelog.

File: src/token/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo
Comment (ericnordelo) [inline src/token/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo:10]: Do we need this import?

File: src/token/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo
Comment (ericnordelo) [inline src/token/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo:116]: Should we provide this as a hooks implementation (after_update)? Something like a function maybe is enough, and then users can call each extension after_update in the preset Hooks implementation. We need to document it, but I think it will be easy enough to use.

Comment (andrew-fleming) [reply src/token/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo:116]: That's basically the idea behind the design. I went back and forth on `update` vs `before_update`. I do like the idea of having a pattern of extension functions (required in hooks) named `before/after_update`. Will update

File: src/tests/mocks/erc721_enumerable_mocks.cairo
Comment (ericnordelo) [inline src/tests/mocks/erc721_enumerable_mocks.cairo:7]:
```suggestion
```

**Review by ericnordelo — COMMENTED** [review]
Hey Andrew, the PR looks great but I have two "big" suggestions to consider before getting it merged:

1. I feel it is worth waiting until at least the next milestone to release it since we are going to have two good additions to the language: Map and StorageArray, that can help both making the implementation cleaner and avoid breaking right after we release a first version. I don't think we should target being backward compatible for this module, which brings us into suggestion two.

2. Break backward compatibility to use Map and StorageArray, and also remove the camelCase interface. I think at this point we are not targeting contracts that may upgrade from cairo0 to this new version, since they either upgraded already or are not gonna do it.

**Comment by andrew-fleming** [comment]
> 1. I feel it is worth waiting until at least the next milestone to release it since we are going to have two good additions to the language: Map and StorageArray, that can help both making the implementation cleaner and avoid breaking right after we release a first version. I don't think we should target being backward compatible for this module, which brings us into suggestion two.

Mmmmm yeah, I think that's fair. I'll change this to a draft and we can revisit in either the `next` or `after` milestone

> 2. Break backward compatibility to use Map and StorageArray, and also remove the camelCase interface. I think at this point we are not targeting contracts that may upgrade from cairo0 to this new version, since they either upgraded already or are not gonna do it.

Agreed. Will update!

File: packages/token/src/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo
Comment (ericnordelo) [inline packages/token/src/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo:122]: Can't we use self instead of PrivateImpl here?

**Review by ericnordelo — COMMENTED** [review]
Sorry for the delay @andrew-fleming. I reviewed the component and it looks great. I think we can go ahead with this version (we may use Vec in the future if we find it worth it). I didn't review the documentation and tests yet, but that's on my priorities.

File: packages/token/src/tests/erc721/test_erc721_enumerable.cairo
Comment (immrsd) [inline packages/token/src/tests/erc721/test_erc721_enumerable.cairo:261]: Should we call mint/burn in the tests instead of calling `before_update` directly? These calls lead to inconsistent state of the component: enumerable mappings get updated, but the balance of the user does not

File: packages/token/src/tests/erc721/test_erc721_enumerable.cairo
Comment (immrsd) [inline packages/token/src/tests/erc721/test_erc721_enumerable.cairo:540]: It would be helpful to implement helper functions to fetch tokens using ERC721Enumerable functionality and then to assert that the result list of tokens is equal to the expected one

```
fn fetch_tokens_by_owner(owner: ContractAddress) -> Span<u256> {
    let contract = CONTRACT_STATE();
    let tokens_len = contract.balance_of(owner);
    let mut i = 0;
    let mut result = array![];
    while i < tokens_len {
        let token_id = contract.token_of_owner_by_index(owner, i);
        result.append(token_id);
        i += 1;
    };
    result.span()
}

fn fetch_all_tokens() -> Span<u256> {
    let contract = CONTRACT_STATE();
    let tokens_len = contract.total_supply();
    let mut i = 0;
    let mut result = array![];
    while i < tokens_len {
        let token_id = contract.token_by_index(i);
        result.append(token_id);
        i += 1;
    };
    result.span()
}
```

File: packages/token/src/tests/erc721/test_erc721_enumerable.cairo
Comment (immrsd) [inline packages/token/src/tests/erc721/test_erc721_enumerable.cairo:552]:
```suggestion
    let actual_token_list = fetch_tokens_by_owner(owner);
    assert_eq!(actual_token_list, expected_list, "Owner tokens mismatch after update");
```

File: packages/token/src/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo
Comment (andrew-fleming) [reply packages/token/src/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo:160]: Hm I just checked to see the difference and they appear to be the same estimated gas and steps

File: packages/token/src/tests/erc721/test_erc721_enumerable.cairo
Comment (andrew-fleming) [reply packages/token/src/tests/erc721/test_erc721_enumerable.cairo:505]: It's a relic from a previous iteration. Yeah, will remove

File: packages/token/src/tests/erc721/test_erc721_enumerable.cairo
Comment (andrew-fleming) [reply packages/token/src/tests/erc721/test_erc721_enumerable.cairo:261]: These tests are specifically testing the `before_update` fn. The rest of the test suite uses mint and burn. I'll update the `before_update` test names to make this more clear

File: packages/token/src/erc721/extensions/erc721_enumerable/interface.cairo
Comment (andrew-fleming) [reply packages/token/src/erc721/extensions/erc721_enumerable/interface.cairo:14]: In the example, you mean 1 `balance_of` call and 100 storage reads vs 100 `balance_of`/storage reads, right? I think it's an interesting idea for optimization. I'd prefer getting the core functionality merged first before considering non-essential additions. WDYT?

File: packages/token/src/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo
Comment (andrew-fleming) [reply packages/token/src/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo]: Nope, it's updated. `is_zero` should be used

File: packages/token/src/tests/erc721/test_erc721_enumerable.cairo
Comment (andrew-fleming) [reply packages/token/src/tests/erc721/test_erc721_enumerable.cairo]: This would make the assertions a lot cleaner by abstracting out the loops! What I don't like about it though is that it's abstracting out the function call that's being asserted (`token_by_index` and `token_of_owner_by_index`).

One of the loops also needs to read from storage directly for `before_update` tests in order to bypass the out of bounds check. I added comments for clarity. That said, I don't think it's really justified to abstract the assertions further. There's only three loops in the helpers. Let me know what you think and if you disagree

File: docs/modules/ROOT/pages/api/erc721.adoc
Comment (ericnordelo) [inline docs/modules/ROOT/pages/api/erc721.adoc:212]:
```suggestion
==== `[.contract-item-name]#++token_by_index++#++(index: u256) -> u256++` [.item-kind]#external#
```

File: docs/modules/ROOT/pages/api/erc721.adoc
Comment (ericnordelo) [inline docs/modules/ROOT/pages/api/erc721.adoc:219]:
```suggestion
==== `[.contract-item-name]#++token_of_owner_by_index++#++(owner: ContractAddress, index: u256) -> u256++` [.item-kind]#external#
```

File: docs/modules/ROOT/pages/api/erc721.adoc
Comment (ericnordelo) [inline docs/modules/ROOT/pages/api/erc721.adoc:675]:
```suggestion
```

File: docs/modules/ROOT/pages/api/erc721.adoc
Comment (ericnordelo) [inline docs/modules/ROOT/pages/api/erc721.adoc:741]:
```suggestion
use openzeppelin::token::erc721::extensions::ERC721EnumerableComponent;
```

File: docs/modules/ROOT/pages/api/erc721.adoc
Comment (ericnordelo) [inline docs/modules/ROOT/pages/api/erc721.adoc:749]:
```suggestion
WARNING: To properly track token ids, this extension requires that the xref:#ERC721EnumerableComponent-before_update[ERC721EnumerableComponent::before_update] function is called before every transfer, mint, or burn operation.
```

File: packages/token/src/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo
Comment (ericnordelo) [inline packages/token/src/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo:135]: I vote we keep this as part of the InternalImpl and let the PrivateImpl refactor to a separate issue. The main reason I lean towards this is that it doesn't look right IMO to document the PrivateImpl in the docsite, because if it is private, why are we documenting it? But at the same time, it is not private, so we probably should document it. 

I think for the sake of not delaying this, we can stay with InternalImpl and design how PrivateImpl would look like in a different issue.

File: packages/token/src/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo
Comment (ericnordelo) [reply packages/token/src/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo:157]: This should be compiler optimized.

File: packages/token/src/tests/erc721.cairo
Comment (ericnordelo) [inline packages/token/src/tests/erc721.cairo:4]:
```suggestion
```

File: packages/token/src/tests/erc721.cairo
Comment (ericnordelo) [inline packages/token/src/tests/erc721.cairo:6]:
```suggestion
```

File: packages/token/src/tests/mocks/erc721_enumerable_mocks.cairo
Comment (ericnordelo) [inline packages/token/src/tests/mocks/erc721_enumerable_mocks.cairo:75]: We can remove this now, right?

File: packages/token/src/tests/mocks/erc721_enumerable_mocks.cairo
Comment (ericnordelo) [inline packages/token/src/tests/mocks/erc721_enumerable_mocks.cairo:165]: Same here.

**Review by ericnordelo — COMMENTED** [review]
Looking great Andrew! Left some comments mostly on the docs.

File: packages/token/src/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo
Comment (andrew-fleming) [reply packages/token/src/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo:135]: Sounds good to me!

File: packages/token/src/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo
Comment (ericnordelo) [inline packages/token/src/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo:114]: In the Solidity version, this is called after the update, not before. Let us refactor this to match it at least for consistency.

File: docs/modules/ROOT/pages/api/erc721.adoc
Comment (ericnordelo) [inline docs/modules/ROOT/pages/api/erc721.adoc:747]: What do you think about adding an example of how the Hool implementation should look like here? Since we don't have this in Wizard or in a preset, I think that could be very helpful.

We may need to remove the WARNING block for that (keeping the content), but I think that's worth.

**Review by ericnordelo — COMMENTED** [review]
We are very close @andrew-fleming. I left a suggestion for consistency.

File: packages/token/src/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo
Comment (andrew-fleming) [reply packages/token/src/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo:114]: You are correct; however, this will break the design of the current hooks implementation. In the Solidity version, the extension overrides `_update` and first calls `super._update` which returns the previous owner. We can't do this because we're hooking into `update` and there's no `from` param (we're getting `from` from `_owner_of`). As an example, minting a token will not update the total supply because the previous owner will be the token recipient and not the zero address

We can consider a bigger refactor with erc721 if it makes sense to add `from` to `update`

Note also that prior to v5.0.0 in the solidity lib, this extension used `_beforeTokenTransfer`

File: docs/modules/ROOT/pages/api/erc721.adoc
Comment (andrew-fleming) [reply docs/modules/ROOT/pages/api/erc721.adoc:747]: Good idea!

File: packages/token/src/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo
Comment (ericnordelo) [reply packages/token/src/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo:114]: Got it. As you said is better to leave it this way. No need to refactor because of this.

**Review by ericnordelo — APPROVED** [review]
Great job! LGTM.

File: packages/token/src/erc721/extensions/erc721_enumerable/interface.cairo
Comment (immrsd) [reply packages/token/src/erc721/extensions/erc721_enumerable/interface.cairo:14]: Actually, I don't see why 100 `balance_of` could be necessary. We assume balance doesn't change while the list of owned token is being built, so we can safely call `balance_of` only once at the beginning

I was talking about an optimisation on the Frontend side. With 100 tokens it's going to take 101 RPC calls and with a helper function it will take only a single one

I agree it doesn't have to be part of this PR thought. WDYT about adding an additional helper read function for building a list of all tokens owned by a user later?

Comment (andrew-fleming) [reply packages/token/src/erc721/extensions/erc721_enumerable/interface.cairo:14]: > I was talking about an optimisation on the Frontend side... I agree it doesn't have to be part of this PR thought. WDYT about adding an additional helper read function for building a list of all tokens owned by a user later?

For sure! I don't see a strong reason not to, but this should be considered low priority at least for now

File: packages/token/src/tests/erc721/test_erc721_enumerable.cairo
Comment (immrsd) [inline packages/token/src/tests/erc721/test_erc721_enumerable.cairo:511]:
```suggestion
    let state = @COMPONENT_STATE();
    let contract_state = @CONTRACT_STATE();
```

File: packages/token/src/tests/erc721/test_erc721_enumerable.cairo
Comment (immrsd) [inline packages/token/src/tests/erc721/test_erc721_enumerable.cairo:528]:
```suggestion
    let state = @COMPONENT_STATE();
```

File: packages/token/src/tests/mocks/erc721_enumerable_mocks.cairo
Comment (immrsd) [inline packages/token/src/tests/mocks/erc721_enumerable_mocks.cairo:59]: Could you check if `self.get_contract_mut().erc721_enumerable.before_update(to, token_id);` will do the job?

**Review by immrsd — COMMENTED** [review]
Looks good

File: packages/token/src/tests/erc721/test_erc721_enumerable.cairo
Comment (immrsd) [reply packages/token/src/tests/erc721/test_erc721_enumerable.cairo]: I checked the new assertion helpers and completely happy with the implementation 👍

**Review by immrsd — APPROVED** [review]

File: packages/token/src/tests/mocks/erc721_enumerable_mocks.cairo
Comment (andrew-fleming) [reply packages/token/src/tests/mocks/erc721_enumerable_mocks.cairo:59]: The compiler yells that the ref arg must be a variable. Keeping it in one line would be much better though

```
error: ref argument must be a variable.

            self.get_contract_mut().erc721_enumerable.before_update(to, token_id);
            ^***************************************^
```

**Pull request merged**
